feat: Add `ignoreOnInitialization` option to no-shadow rule (#14963)
eslint/eslint: docs/rules/no-shadow.md

});
```

### ignoreOnInitialization

> The `ignoreOnInitialization` option is `false` by default.
> If it is `true`, the rule prevents reporting variables on initialization statements, the shadowed variable must be on the left side and the shadowing variable must be on the right side.
> The option reports also the shadowing variables that are inside IIFE.

Reviewer: ```suggestion
The `ignoreOnInitialization` option is `false` by default. If it is `true`, it prevents reporting shadowing of variables in their initializers when the shadowed variable is presumably still uninitialized.

The shadowed variable must be on the left side. The shadowing variable must be on the right side and declared in a callback function or in an IIFE.
```

> Examples of **incorrect** code for the default `{ "ignoreOnInitialization": "true" }` option:

Reviewer: ```suggestion
Examples of **incorrect** code for the `{ "ignoreOnInitialization": "true" }` option:
```

`{ "ignoreOnInitialization": "true" }` isn't by default.

```js
/*eslint no-shadow: ["error", { "ignoreOnInitialization": true }]*/

var x = x => x;
```

> Because the shadowing variable x will shadow an initialized shadowed variable x.

Reviewer: ```suggestion
Because the shadowing variable `x` will shadow the already initialized shadowed variable `x`.
```

Examples of **correct** code for the `{ "ignoreOnInitialization": true }` option:

```js
/*eslint no-shadow: ["error", { "ignoreOnInitialization": true }]*/

> var x = foo(x => x)
> 
> var x = (x => x)()

Reviewer: ```suggestion
var x = foo(x => x)

var y = (y => y)()
```

To avoid confusion with redeclarations.

```

> Because the shadowing variable x will shadow an uninitialized shadowed variable x.

Reviewer: ```suggestion
The rationale for callback functions is the assumption that they will be called during the initialization, so that at the time when the shadowing variable will be used, the shadowed variable has not yet been initialized.
```
